src/Mvc/Mvc.Core/src/ModelBinding/Metadata/DefaultBindingMetadataProvider.cs

```diff
@@ -72,6 +73,77 @@ public void CreateBindingMetadata(BindingMetadataProviderContext context)
                 context.BindingMetadata.IsBindingAllowed = bindingBehavior.Behavior != BindingBehavior.Never;
                 context.BindingMetadata.IsBindingRequired = bindingBehavior.Behavior == BindingBehavior.Required;
             }
+
+            if (GetBoundConstructor(context.Key.ModelType) is ConstructorInfo constructorInfo)
+            {
+                context.BindingMetadata.BoundConstructor = constructorInfo;
+            }
+        }
+
+        internal static ConstructorInfo GetBoundConstructor(Type type)
+        {
+            if (type.IsAbstract || type.IsValueType || type.IsInterface)
+            {
+                return null;
+            }
+
+            var constructors = type.GetConstructors();
+            if (constructors.Length == 0)
+            {
+                return null;
+            }
+
+            var parameterlessConstructor = constructors.FirstOrDefault(f => f.GetParameters().Length == 0);
+            return GetRecordTypeConstructor(type, constructors) ?? parameterlessConstructor;
+        }
+
+        private static ConstructorInfo GetRecordTypeConstructor(Type type, ConstructorInfo[] constructors)
+        {
+            if (!IsRecordType(type))
+            {
+                return null;
+            }
+
+            // For record types, we will support binding and validating the primary constructor.
+            // There isn't metadata to identify a primary constructor. We require that at most one constructor is defined on the type,
+            // and that every parameter on the constructor has a corresponding property.
+
+            if (constructors.Length > 1)
+            {
+                return null;
+            }
+
+            var constructor = constructors[0];
+
+            var parameters = constructor.GetParameters();
+            var properties = PropertyHelper.GetVisibleProperties(type);
+
+            for (var i = 0; i < parameters.Length; i++)
+            {
+                var parameter = parameters[i];
+                var mappedProperties = properties.Where(property =>
+                    string.Equals(property.Name, parameter.Name, StringComparison.OrdinalIgnoreCase) &&
+                    property.Property.PropertyType == parameter.ParameterType)
+                    .ToList();
+
+                if (mappedProperties.Count == 0)
+                {
+                    // No property found, this is not a primary constructor.
+                    return null;
+                }
+                else if (mappedProperties.Count > 1)
+                {
+                    // More than one property found that maps to a constructor parameter. This is ambigious.
+                    return null;
+                }
+            }
+
+            return constructor;
+
+            static bool IsRecordType(Type type)
+            {
+                return type.GetProperty("EqualityContract", BindingFlags.NonPublic | BindingFlags.Instance) != null;
```

Comment:
Someone could have `EqualityContract` property on a type.

Is there no attribute on record type classes to say it is a record type?

Reply:
Not that I can tell. I have https://github.com/dotnet/aspnetcore/issues/23975 to follow up with the compiler team. What we really want is the ability to identify the primary constructor. If we had a reliable way to identify that, we could remove this check, unambiguously always say that it gets bound \ validated and remove the 1-constructor limit.

Reply:
A "primary" constructor isn't required as far as I can tell. What documentation am I missing❔

And, I'd say we need to unambiguously identify record types _before_ trying to find how best to bind it.

Reply:
> A "primary" constructor isn't required as far as I can tell. What documentation am I missing❔

Nevermind. If we need to bind a constructor, the primary constructor will exist.

However it's a fairly significant deficit that we can't find it or even detect `record` types reliably and have this single constructor requirement.